database/src/service/read.rs

```diff
@@ -230,55 +241,81 @@ fn map_request(
 
 /// [`ReadRequest::BlockExtendedHeader`].
 #[inline]
-fn block_extended_header(env: &Arc<ConcreteEnv>, block_height: BlockHeight) -> ResponseResult {
-    todo!()
+fn block_extended_header(env: &ConcreteEnv, block_height: BlockHeight) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let tx_ro = env_inner.tx_ro()?;
+    let tables = env_inner.open_tables(&tx_ro)?;
+
+    Ok(Response::BlockExtendedHeader(
+        get_block_extended_header_from_height(&block_height, &tables)?,
+    ))
 }
 
 /// [`ReadRequest::BlockHash`].
 #[inline]
-fn block_hash(env: &Arc<ConcreteEnv>, block_height: BlockHeight) -> ResponseResult {
-    todo!()
+fn block_hash(env: &ConcreteEnv, block_height: BlockHeight) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let tx_ro = env_inner.tx_ro()?;
+    let table_block_infos = env_inner.open_db_ro::<BlockInfos>(&tx_ro)?;
+
+    Ok(Response::BlockHash(
+        get_block_info(&block_height, &table_block_infos)?.block_hash,
+    ))
 }
 
 /// [`ReadRequest::BlockExtendedHeaderInRange`].
 #[inline]
 fn block_extended_header_in_range(
-    env: &Arc<ConcreteEnv>,
+    env: &ConcreteEnv,
     range: std::ops::Range<BlockHeight>,
 ) -> ResponseResult {
-    todo!()
+    let env_inner = env.env_inner();
+
+    // This iterator will early return as `Err` if there's even 1 error.
+    let vec = {
+        range
+            .into_par_iter()
+            .map(|block_height| {
+                let tx_ro = env_inner.tx_ro()?;
+                let tables = env_inner.open_tables(&tx_ro)?;
+                get_block_extended_header_from_height(&block_height, &tables)
+            })
+            .collect::<Result<Vec<ExtendedBlockHeader>, RuntimeError>>()?
+    };
+
+    Ok(Response::BlockExtendedHeaderInRange(vec))
```

Comment:
@Boog900 Re: https://github.com/Cuprate/cuprate/pull/102#discussion_r1554580888

LMDB docs:
> A thread can only use one transaction at a time, plus any child transactions. Each transaction belongs to one thread.

And `heed` (issue 20):

> A thread is a single unit of execution. So is a transaction. It is nonsensical to associate them other than 1:1.

So we'll be opening a read transaction + table(s) per reader thread.

And although `redb`'s is `Send + Sync`, it seems like the read transaction + table per thread model is correct there too:

https://github.com/cberner/redb/blob/0d0e11ea901a3fe8a7ec4ac0b2b4aee121d11461/benches/lmdb_benchmark.rs#L220-L225

Reply:
If we open RO LMDB txs with `MDB_NOTLS` we can impl Send for our read only txs.

We can then use a [thread local](https://docs.rs/thread_local/latest/src/thread_local/lib.rs.html#103-111) which will hold the tables, so we don't need to keep opening txs and tables.

Reply:
Are you sure 1 TX per thread is not how reads scale? Issue 20 seems to answer the `MDB_NOTLS` question:
> yes but a txn can only be used by one thread at a time, regardless. concurrent use of a single txn from multiple threads will cause memory corruption.

I believe it's mostly for moving the transaction, not sharing a reference across many threads.

This does break read atomicity though, AFAIK this is prevented in `monerod` with a spinlock, we could use a `RwLock` (only inside `service`) that syncs the reader-pool + 1 writer.

Reply:
> Are you sure 1 TX per thread is not how reads scale

We would still have 1 tx per thread but currently this is doing 1 tx per _block height_, using a thread local allows us to keep 1 tx per thread.

> This does break read atomicity though, AFAIK this is prevented in monerod with a spinlock, we could use a RwLock (only inside service) that syncs the reader-pool + 1 writer.

I am not really worried about this too much, the syncer will have a consistent view as it waits on writes so the worse event that we could get is an RPC (or other interface) request during a reorg when one tx sees one chain and the other sees another, the chance of this happening though are very low.

What we could do to combat this:

- Keep an atomic counter that gets incremented on a reorg (no request to do this yet)
- When a tx is added to the thread local record the atomics value with the tx. i.e (Tx, u64)
- after retrieving all values loop over the thread local and check all the txs have the same value.
- If they differ, try again.

It doesn't matter if they differ with the current atomic value just that they all agree on the state.

This does still allow for 2 txs to disagree on the top block but I can't see an issue with this.

Reply:
> We would still have 1 tx per thread but currently this is doing 1 tx per block height, using a thread local allows us to keep 1 tx per thread.

Ahhh okay I see the mistake. `heed` has a `read-txn-no-tls` feature for `MDB_NOTLS` I'll enable.

> the syncer will have a consistent view as it waits on writes

IMO the correctness shouldn't depend on external code using it correctly.

> the worse event that we could get is an RPC (or other interface) request during a reorg when one tx sees one chain and the other sees another

Wouldn't a normal `add_block()` call be bad?
```
Reader 1 (same request as reader 2)
   |
   v      Writer
 tx_ro      |
   |        v        Reader 2
   |  tx_rw + commit   |
   |                   v
    \                tx_ro <- different state than reader 1
      \________________|
            |
collecting from different database states
```

> What we could do to combat this:

Do you think a counter + retries would allow more reader throughput than a `RwLock`? Is it worth taking away simplicity? Each function will now need `loop { do(); maybe_break }` style code instead of `lock(); do()`.

Could we use `RwLock` for now and implement a counter later if there's bottlenecks?

Reply:
> Wouldn't a normal add_block() call be bad?

An add block only adds new data and does not change old data, the worst that can happen is one tx does not see the new data and the request fails, which would have happened anyway, even with locking, if that same request came a couple seconds earlier.

> Do you think a counter + retries would allow more reader throughput than a RwLock

Yes I do, I don't know how significant the difference would be but I don't really want to add a lock for this.

> Each function will now need loop { do(); maybe_break }

Or you could make it recurse, which would be fine as this event is extremely unlikely.

> Could we use RwLock for now and implement a counter later if there's bottlenecks?

I would rather just not handle this now than add a lock.